src/main/java/frc/robot/swerve/SwerveSubsystem.java

```diff
@@ -371,144 +386,235 @@ public Command driveOpenLoopFieldRelative(Supplier<ChassisSpeeds> speeds) {
+                    : FieldUtils.RED_REEF_CENTER)
+            .getNorm()
+        < toleranceMeters;
+  }
+
+  public Command autoAimToL23(BooleanSupplier leftHanded) {
+    return translateToPose(
+        () -> {
+          Twist2d twist = getVelocityFieldRelative().toTwist2d(0.3);
+          Transform2d twistTransform =
+              new Transform2d(twist.dx, twist.dy, Rotation2d.fromRadians(twist.dtheta));
+          return FieldUtils.CoralTargets.getHandedClosestTargetL23(
+              getPose().plus(twistTransform), leftHanded.getAsBoolean());
+        });
   }
 
   public boolean nearL23() {
-    // TODO
-    return true;
+    return isInAutoAimTolerance(FieldUtils.CoralTargets.getClosestTargetL23(getPose()));
   }
 
-  public Command autoAimToL4() {
-    // TODO
-    return Commands.none();
+  public Command autoAimToL4(BooleanSupplier leftHanded) {
+    return translateToPose(
+        () -> {
+          Twist2d twist = getVelocityFieldRelative().toTwist2d(0.3);
+          Transform2d twistTransform =
+              new Transform2d(twist.dx, twist.dy, Rotation2d.fromRadians(twist.dtheta));
+          return FieldUtils.CoralTargets.getHandedClosestTargetL4(
+              getPose().plus(twistTransform), leftHanded.getAsBoolean());
+        });
   }
 
   public boolean nearL4() {
-    // TODO
-    return true;
+    return isInAutoAimTolerance(FieldUtils.CoralTargets.getClosestTargetL4(getPose()));
   }
 
-  public Command autoAimToOffsetAlgae() {
-    // TODO
-    return Commands.none();
+  public Command autoAimToOffsetAlgaePose() {
+    return translateToPose(
+        () ->
+            FieldUtils.AlgaeIntakeTargets.getOffsetLocation(
+                FieldUtils.AlgaeIntakeTargets.getClosestTargetPose(getPose())));
   }
 
   public boolean nearIntakeAlgaeOffsetPose() {
-    // TODO
-    return true;
+    return isInAutoAimTolerance(
+        FieldUtils.AlgaeIntakeTargets.getOffsetLocation(
+            FieldUtils.AlgaeIntakeTargets.getClosestTargetPose(getPose())));
   }
 
   public Command approachAlgae() {
-    // TODO
-    return Commands.none();
+    return driveClosedLoopRobotRelative(
+        () -> {
+          AutoAim.resetPIDControllers(getPose(), getVelocityFieldRelative());
+          ChassisSpeeds calculatedSpeedsRobotRelative =
+              ChassisSpeeds.fromFieldRelativeSpeeds(
+                  AutoAim.calculateSpeeds(
+                      getPose(), FieldUtils.AlgaeIntakeTargets.getClosestTargetPose(getPose())),
+                  getRotation());
+          calculatedSpeedsRobotRelative.vxMetersPerSecond =
+              isInAutoAimTolerance(FieldUtils.AlgaeIntakeTargets.getClosestTargetPose(getPose()))
+                  ? AutoAim.ALGAE_APPROACH_SPEED_METERS_PER_SECOND
+                  : 0.0;
+          // IDK why we have to do this but it was in reefscape so...
+          calculatedSpeedsRobotRelative.vyMetersPerSecond =
+              -calculatedSpeedsRobotRelative.vyMetersPerSecond;
```

Comment:
```suggestion
                    calculatedSpeedsRobotRelative.vyMetersPerSecond *= -1;

```
minus signs are hard to read imo